# SOLUTIONS AND EXPLANATIONS FOR LESSON 1: MATRIX FOR AI (20/6/2026) HOMEWORK - USING TEXTS

## PREPARATION FOR BOTH EXERCISES: TRANSFORM TEXTS TO VECTORS AND MATRIXES

Import numpy lib and create a list of 8 sentences to transform to vectors

In [2]:
import numpy as np
sentence = ["i love you", 
            "this house is big",
            "you are beautiful",
            "that will do",
            "you are welcome", 
            "how are you", 
            "what is your name", 
            "i play badminton"]

extract each separate word from every sentence and then sort them in alphabet order

In [3]:
word = []
for s in sentence:
    for w in s.lower().split():
        word.append(w)
word.sort()
print(word)

['are', 'are', 'are', 'badminton', 'beautiful', 'big', 'do', 'house', 'how', 'i', 'i', 'is', 'is', 'love', 'name', 'play', 'that', 'this', 'welcome', 'what', 'will', 'you', 'you', 'you', 'you', 'your']


building a function to transform these words to vectors

In [4]:
def to_vectors(s):
    v = np.zeros(len(word)) # an array all 0 with length of word list above
    for w in s.lower().split():
        v[word.index(w)] += 1 
    return v
matrix = np.array([to_vectors(s) for s in sentence], dtype=int)
print(matrix)
print(matrix.shape)

[[0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0]
 [0 0 0 0 0 1 0 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0]
 [1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0]
 [0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0]
 [1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0]
 [1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0 1]
 [0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0]]
(8, 26)


Here, 8 sentences are separately transformed to 8 vectors (each with 26 dims or 26 words in total) which makes a 8x26 matrix.
- 8 rows are 8 sentences whose "1" stands at the index corresponds to a word in the sorted word list above
- Each column represents the frequency that each word appears in each sentence. for ex, the first column represents for word[0] ('are') that 'are' appears in sentence 3,5 and 6 one time.

## EXERCISE 1: CALCULATE THE SIMILARITY

### Basic operation: Mean-centering

We take the mean of each column

In [14]:
col_mean = np.mean(matrix, axis=0, keepdims=True) # axis=0 means that we are taking average by columns (=1 is for rows)
print(col_mean)
print(col_mean.shape)

[[0.375 0.    0.    0.125 0.125 0.125 0.125 0.125 0.125 0.25  0.    0.25
  0.    0.125 0.125 0.125 0.125 0.125 0.125 0.125 0.125 0.5   0.    0.
  0.    0.125]]
(1, 26)


first column average = 3/8 = 0.375. Here we keep the original dims with 26 columns

Now, mean centering

In [16]:
centered_matrix = matrix - col_mean
print(centered_matrix)
print(centered_matrix.shape)

[[-0.375  0.     0.    -0.125 -0.125 -0.125 -0.125 -0.125 -0.125  0.75
   0.    -0.25   0.     0.875 -0.125 -0.125 -0.125 -0.125 -0.125 -0.125
  -0.125  0.5    0.     0.     0.    -0.125]
 [-0.375  0.     0.    -0.125 -0.125  0.875 -0.125  0.875 -0.125 -0.25
   0.     0.75   0.    -0.125 -0.125 -0.125 -0.125  0.875 -0.125 -0.125
  -0.125 -0.5    0.     0.     0.    -0.125]
 [ 0.625  0.     0.    -0.125  0.875 -0.125 -0.125 -0.125 -0.125 -0.25
   0.    -0.25   0.    -0.125 -0.125 -0.125 -0.125 -0.125 -0.125 -0.125
  -0.125  0.5    0.     0.     0.    -0.125]
 [-0.375  0.     0.    -0.125 -0.125 -0.125  0.875 -0.125 -0.125 -0.25
   0.    -0.25   0.    -0.125 -0.125 -0.125  0.875 -0.125 -0.125 -0.125
   0.875 -0.5    0.     0.     0.    -0.125]
 [ 0.625  0.     0.    -0.125 -0.125 -0.125 -0.125 -0.125 -0.125 -0.25
   0.    -0.25   0.    -0.125 -0.125 -0.125 -0.125 -0.125  0.875 -0.125
  -0.125  0.5    0.     0.     0.    -0.125]
 [ 0.625  0.     0.    -0.125 -0.125 -0.125 -0.125 -0.125  0

Numpy will automatically copy the col_mean matrix above to 7 more rows to make a 8x26 matrix (same shape as matrix). This is called broadcasting in Numpy.  
Subtracting each entry with the average will make each column's mean = 0.

In [17]:
column_mean = np.mean(centered_matrix, axis=0, keepdims=True)
print(column_mean)

[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0.]]


### Cosine similarity

is basically comparing the value of cosine between 2 vectors:
- ~1 means they are highly related

In [18]:
def cosine_similarity(X, Y = None):
    if Y is None:
        Y = X
    
    Xn = X / np.linalg.norm(X, axis=1) # divided by the length (norm) of itself or Vector Normalization
    Yn = Y / np.linalg.norm(Y, axis=1)

    return Xn @ Yn.T

This function corresponds to the formula: $cos = \frac{u \times v}{||u|| \times ||v||}$

In [29]:
# Calculate with sentence 1 and 2
print(f'cosine similarity of sentence 1 and 2 is: {cosine_similarity(matrix[0:1], matrix[1:2])}')

# Calculate with sentence 1 and 8
print(f'cosine similarity of sentence 1 and 8 is: {cosine_similarity(matrix[0:1], matrix[7:8])}')

# Calculate with sentence 3 and 5
print(f'cosine similarity of sentence 3 and 5 is: {cosine_similarity(matrix[2:3], matrix[4:5])}')


cosine similarity of sentence 1 and 2 is: [[0.]]
cosine similarity of sentence 1 and 8 is: [[0.33333333]]
cosine similarity of sentence 3 and 5 is: [[0.66666667]]


sentence 1 and 2 have 0/3 identical words. Cos = 0 means 2 vectors are perpendicular and are not related  
whereas 1 and 8 have 1/3 and 3 and 5 have 2/3

### Word Embedding - Semantic research

The to_vector method above is just a way to vectorize a sentence to count the appearance of words. It does not account for the semantic or meaning of them.  
For example, 'I go by bike' and 'I ride a bicycle' would end up having a very ~0 cosine value as the function doesn't know that bike and bicycle are synonyms or go by and ride is almost the same.


This is where Semantic search is introduced.  
First, engineers create a vector that has 100-768 dims (most LLMs now choose 768) representing for a separate word. Each dim is a randomly generated number which show a property of that word.  
Then the model will go through 2 games: masked language modelling and next sentence prediction to adjust those numbers so that each word will have its own set and the relations between vocabs are precise. For ex, 'cat' and 'dog' will be closed not 'cat' and 'car'.  
Finally, when users input, the model will use cosine similarity to find the top-related words in terms of context and meaning.